odoo_backup_sh/models/odoo_backup_sh.py

```diff
@@ -16,67 +21,289 @@
+                    needed_backup_dts = [last_backup_dt]
+                    backup_config_time_frames = []
+                    if backup_config.hourly > 1:
+                        last_backup_dt_hour = last_backup_dt.replace(minute=0, second=0)
+                        for hour in range(1, backup_config.hourly):
+                            backup_config_time_frames.append([last_backup_dt_hour - timedelta(hours=hour),
+                                                              last_backup_dt_hour - timedelta(hours=hour-1)])
+                    if backup_config.daily > 1:
+                        last_backup_dt_day = last_backup_dt.replace(hour=0, minute=0, second=0)
+                        for day in range(1, backup_config.daily):
+                            backup_config_time_frames.append([last_backup_dt_day - timedelta(days=day),
+                                                              last_backup_dt_day - timedelta(days=day-1)])
+                    if backup_config.weekly > 1:
+                        last_backup_dt_day = last_backup_dt.replace(hour=0, minute=0, second=0)
+                        last_backup_dt_week = last_backup_dt_day - timedelta(days=last_backup_dt_day.weekday())
+                        for week in range(1, backup_config.weekly):
+                            backup_config_time_frames.append([last_backup_dt_week - timedelta(weeks=week),
+                                                              last_backup_dt_week - timedelta(weeks=week-1)])
+                    if backup_config.monthly > 1:
+                        last_backup_dt_month = last_backup_dt.replace(day=0, hour=0, minute=0, second=0)
+                        for month in range(1, backup_config.monthly):
+                            backup_config_time_frames.append([last_backup_dt_month - relativedelta(months=month),
+                                                              last_backup_dt_month - relativedelta(months=month-1)])
+                    if backup_config.yearly > 1:
+                        last_backup_dt_year = last_backup_dt.replace(month=0, day=0, hour=0, minute=0, second=0)
+                        for year in range(1, backup_config.yearly):
+                            backup_config_time_frames.append([last_backup_dt_year - relativedelta(years=year),
+                                                              last_backup_dt_year - relativedelta(years=year-1)])
+                    while backup_config_time_frames:
+                        frame = backup_config_time_frames.pop()
+                        if any(frame[0] < backup_dt < frame[1] for backup_dt in needed_backup_dts):
+                            continue
+                        for backup_dt in sorted_backup_dts:
+                            if frame[0] < backup_dt < frame[1]:
+                                needed_backup_dts.append(backup_dt)
+                                sorted_backup_dts.remove(backup_dt)
```

Comment:
Computation of needed_backup_dts and sorted_backup_dts has to be coverted, sorry

Reply:
ok, I will separate this code and create tests for it.